src/openrct2/scripting/ScriptEngine.cpp

```diff
@@ -1645,42 +1649,47 @@ void ScriptEngine::UpdateIntervals()
     {
         // timestamp has wrapped, subtract all intervals by the remaining amount before wrap
         auto delta = static_cast<int64_t>(std::numeric_limits<uint32_t>::max() - _lastIntervalTimestamp);
-        for (auto& interval : _intervals)
+        for (auto& entry : _intervals)
         {
-            if (interval.IsValid())
-            {
-                interval.LastTimestamp = -delta;
-            }
+            auto& interval = entry.second;
+            interval.LastTimestamp = -delta;
         }
     }
     _lastIntervalTimestamp = timestamp;
 
-    for (auto& interval : _intervals)
+    // This loop needs to account for removal and insertions during iteration.
+    for (auto it = _intervals.begin(), itNext = it; it != _intervals.end(); it = itNext)
     {
-        if (interval.IsValid())
+        itNext++;
+
+        auto& interval = it->second;
+
+        if (timestamp >= interval.LastTimestamp + interval.Delay)
         {
-            if (timestamp >= interval.LastTimestamp + interval.Delay)
-            {
-                ExecutePluginCall(interval.Owner, interval.Callback, {}, false);
+            ExecutePluginCall(interval.Owner, interval.Callback, {}, false);
 
-                interval.LastTimestamp = timestamp;
-                if (!interval.Repeat)
-                {
-                    RemoveInterval(nullptr, interval.Handle);
-                }
+            interval.LastTimestamp = timestamp;
+            if (!interval.Repeat)
+            {
+                _intervals.erase(it);
```

Comment:
Doesn't `erase` return the next iterator after the one removed?
I think the code could be just:

```c++
    for (auto it = _intervals.begin(), it != _intervals.end();)
    {
        auto& interval = it->second;
        if (timestamp >= interval.LastTimestamp + interval.Delay)
        {
            ExecutePluginCall(interval.Owner, interval.Callback, {}, false);
            interval.LastTimestamp = timestamp;
            if (!interval.Repeat)
            {
                it = _intervals.erase(it);
            }
            else
            {
                ++it;
            }
        }
    }
```

Reply:
That will cause an infinite loop, I opted to just use a secondary variable so it doesn't need to increment the iterator in two places.

Reply:
It wouldn't increment the iterator if it hasn't reached the delay you only accounted for the nested condition which would require another else for `if (timestamp >= interval.LastTimestamp + interval.Delay)`. I left it as is for now, iterators in c++ are quite annoying at times.